modules/mbedtls/CMakeLists.txt

```diff
@@ -25,6 +25,10 @@ if(CONFIG_MBEDTLS_BUILTIN)
 
   zephyr_library_sources_ifdef(CONFIG_MBEDTLS_SHELL shell.c)
 
+  # mbedtls configs sometimes have unused vars and funcs
+  zephyr_library_compile_options(-Wno-unused-variable)
+  zephyr_library_compile_options(-Wno-unused-function)
+
```

Comment:
Just wondering if that should be done in the module itself, otherwise this will affect zephyr files.

Just for curiosity, how are you getting warnings ? Our test is not showing anything.

Reply:
These are the build errors that i see

```
-o modules/mbedtl
s/CMakeFiles/modules__mbedtls.dir/home/dave/projects/nuvia/modules/crypto/mbedtls/library/ssl_msg.c.obj -c /home/dave/projects/nuvia/modules/crypto/mbedtls/library/ssl_msg.c
/home/dave/projects/nuvia/modules/crypto/mbedtls/library/ssl_msg.c: In function 'mbedtls_ssl_encrypt_buf':
/home/dave/projects/nuvia/modules/crypto/mbedtls/library/ssl_msg.c:529:12: error: unused variable 'add_data_len' [-Werror=unused-variable]
  529 |     size_t add_data_len;
      |            ^~~~~~~~~~~~
/home/dave/projects/nuvia/modules/crypto/mbedtls/library/ssl_msg.c:528:19: error: unused variable 'add_data' [-Werror=unused-variable]
  528 |     unsigned char add_data[13 + 1 + MBEDTLS_SSL_CID_OUT_LEN_MAX ];
      |                   ^~~~~~~~
/home/dave/projects/nuvia/modules/crypto/mbedtls/library/ssl_msg.c: In function 'mbedtls_ssl_decrypt_buf':
/home/dave/projects/nuvia/modules/crypto/mbedtls/library/ssl_msg.c:1007:12: error: unused variable 'add_data_len' [-Werror=unused-variable]
 1007 |     size_t add_data_len;
      |            ^~~~~~~~~~~~
/home/dave/projects/nuvia/modules/crypto/mbedtls/library/ssl_msg.c:1006:19: error: unused variable 'add_data' [-Werror=unused-variable]
 1006 |     unsigned char add_data[13 + 1 + MBEDTLS_SSL_CID_IN_LEN_MAX ];
      |                   ^~~~~~~~
/home/dave/projects/nuvia/modules/crypto/mbedtls/library/ssl_msg.c:1001:9: error: unused variable 'ret' [-Werror=unused-variable]
 1001 |     int ret, auth_done = 0;
      |         ^~~
/home/dave/projects/nuvia/modules/crypto/mbedtls/library/ssl_msg.c:999:12: error: unused variable 'olen' [-Werror=unused-variable]
  999 |     size_t olen;
      |            ^~~~
At top level:
/home/dave/projects/nuvia/modules/crypto/mbedtls/library/ssl_msg.c:382:13: error: 'ssl_extract_add_data_from_record' defined but not used [-Werror=unused-function]
  382 | static void ssl_extract_add_data_from_record( unsigned char* add_data,
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
ninja: build stopped: subcommand failed.
```

I fully appreciate that this patch has the effect of masking this badness, rather than root cause fixing it.
I did look into addressing the root cause of the issue in upstream mbedtls but it was not clear to me how to test such a change and I could also not find out how upstream mbedtls changes make it back into zephyr.

Reply:
The problem in doing it this place is that we end up applying this flag to our code too. But that is just for these few files. Not a blocker for me.
Regarding the mbedTLS upstream process, once we have a mbedTLS release with these changes we can update our repository to use that release. But this may take a long time since it depends on mbedTLS release process.

Reply:
Thanks for the clarification on the mbedTLS release/update in Zephyr process.

Reply:
please note that if this is only seen on a single / few files, then it is possible to add such flags directly to the file.

Like this:
```
set_source_files_properties(library/ssl_msg.c PROPERTIES COMPILE_OPTIONS  "-Wno-unused-variable;-Wno-unused-function")
```

Second note, to avoid gcc centric flags and be open to more toolchains, then compiler flags should generally be defined here: https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/compiler/gcc/compiler_flags.cmake and then applied as a generator expression.

Because this particular case is expected to be fixed in newer mbedTLS, and the fact the officially supported toolchains all understand this flag, then applying the flag directly is accepted in this case.

But please make a comment in the CMake code as to when this is expected to be fixed in mbedTLS and the compile flag can be removed again.

Reply:
@tejlmand thanks for the great feedback. I'll go make the change for the single source file no worries. As to your second note about adding a comment to CMake about removal. I'm not sure I have the knowledge to do that. The whole mbedTLS update,  pull into Zephyr at some point thing seems to be very opaque to me. This is the reason for this sticky plaster approach to this MR rather then a root cause fix in mbedTLS. So, whilst I completely agree with your comment I do not know how to address it.

Reply:
a comment in CMake like this would be very nice:
```
# mbedTLS v3.1.0 is having unused variables and functions in <file-list>
# To avoid compilation warnings, which are treated as errors in CI, we disable unused variables and functions.
# Please check when mbedTLS is updated to version >v3.1.0 if those flags are still needed.
# If mbedTLS has fixed the mentioned issue, then please remove the flags.
```

Reply:
awesome, thanks